### ISSUE TITLE
Swordfish reads the whole row group instead of streaming chunks?

### ISSUE THREAD

**Issue by sagiahrac** [Eventual-Inc/Daft#3125]
### Describe the bug

Reading a 2GB parquet file with a single row group:
![image](https://github.com/user-attachments/assets/dfa36531-9638-405d-82d0-77a52ba8a4a9)

Reading a 1-row parquet file:
![image](https://github.com/user-attachments/assets/28f9fc87-4006-4520-8b43-3ca71249d31d)

### To Reproduce

Write the parquets (couldn't keep it into a single row group with daft/pyarrow):

```python
import fastparquet
import numpy as np
import pandas as pd
from pyarrow import parquet as pq

GIB = 1 << 30


def write_parquet(dst: str, n_rows: int):
    df = pd.DataFrame(
        {"x": np.random.randint(low=0, high=256, size=n_rows, dtype=np.uint8)}
    )
    fastparquet.write(
        filename=dst, data=df, row_group_offsets=n_rows + 1, compression="SNAPPY"
    )

    assert pq.read_metadata(dst).num_row_groups == 1


if __name__ == "__main__":
    big_parquet = write_parquet(dst="big.parquet", n_rows=GIB)
    small_parquet = write_parquet(dst="small.parquet", n_rows=1)
```

Read (in a different process to avoid GC cleaning the written parquets in the middle of the benchmark):
```python
import daft


if __name__ == "__main__":
    daft.set_execution_config(enable_native_executor=True, default_morsel_size=1)
    df = daft.read_parquet("big.parquet")
    assert df.num_partitions() == 1
    
    df_iter = df.iter_rows(results_buffer_size=1)
    next(df_iter)
```

### Expected behavior

I would expect the memory usage for reading the first row to be significantly less the memory required for materializing the whole dataframe, while using morsel size of 1. Any thoughts?

### Component(s)

Parquet

**Comment by colin-ho** [comment]
Couple things here:
1. Morsel size is currently only applied to operators, e.g. filters, url_downloads, not reads.
2. read_parquet reads in 1 row group at a time into memory. This is why memory usage is high for "big.parquet", which only has 1 row group. Only upon reading the row group, it will deserialize and stream the row group in chunks of 128kb. This is why `iter_rows` will be faster than `.collect()`. 
3. Regarding:
> when saving the parquet file with multiple row groups (I tested with 22), reading the first row still consumes a lot of memory. However, if we read just the first row group, the memory usage is much lower.

Memory pressure in swordfish isn't fully fleshed out yet, so nothing is preventing the reader from reading all the row groups. This is something we're currently working on!

**Comment by jaychia** [comment]
> read_parquet reads in 1 row group at a time into memory.

QQ @colin-ho is this something that can be optimized in the future? I.e. some kind of stateful reading that would materialize at page boundaries instead of at rowgroup boundaries.

I imagine this could get complicated with nested types.

**Comment by colin-ho** [comment]
Yeah it can be optimized, we just need to change the reader to be more granular. This is just for local btw, the remote parquet reader already reads at the page level.

**Comment by sagiahrac** [comment]
Thanks for the detailed response! I’ll avoid using Swordfish with larger datasets for now to prevent OOM :)

**Comment by colin-ho** [comment]
Gonna close this for now as the implementation will be quite involved and is not planned.

The way parquet reading works right now is that row groups are loaded into memory, but the deserialization can be performed in configurable chunks, for example via `into_batches` or the `_chunk_size` param. 

Loading ranges from row groups partially is more challenging (and also possibly worse performance) because it will require identifying the appropriate byte ranges, and lead to a lot more random access.
